### ISSUE TITLE
Separate out KHR_collision from KHR_rigid_bodies (and remove triggers)

### ISSUE THREAD

**Issue by bhouston** [eoineoineoin/glTF_Physics#1]
I think that an extension that just specifies collision bodies would be useful outside of KHR_rigid_bodies (which is what I would name this proposal.)

I think that KHR_collisions can be useful for a lot of different things.  Just as VR controller hits, mouse clicks on objects, and touches on objects.  I also think they can be useful in KHR_behavior for detecting overlaps between objects.

I do not think we need the trigger part of this KHR_rigid_bodies, instead KHR_behavior will have overlap detection and it can act on that.  I suggest not having it in trigger in this extension because it doesn't really do anything here by itself.  Instead it is use using KHR_collision to trigger events.

**Comment by javagl** [comment]
Going one step further (and I know, this is a long shot): I think that splitting the _geometry definition_ and its _usage_ could make sense. There already are proposals for "defining geometric objects", and this is related to [a comment that I left on one of these proposals](https://github.com/KhronosGroup/glTF/issues/1986#issuecomment-1023290360).

The current proposal adds the extension object to a `node`. There may be a `mesh` attached to that node. And the extension object itself defines the `collider`. The `collider` may be again a `mesh`, or one of the special `collider` types (like `shere`).

For example, the following node _renders_ a `mesh:0`, and uses a `mesh:1` as its collider:
```
  "nodes" : [
    {
      "mesh" : 0,
      ...
      "extensions" : {
        "MSFT_physics" : {
          "collider" : {
            ...
            "convex" : {
              "mesh" : 1
            }
          }
        }
      },
    },
```
The `mesh:1` can be imagined as an approximation of the rendered mesh. This approximation is supposed to be used for the collision detection. In this case, it is a convex hull, but for other meshes, it may be an even simpler representation, like a `sphere`.

(The fact that the `convex` refers to a `mesh` raises the question: How to ensure that this mesh actually _is_ a convex hull, and not a "real", arbitrarily-shaped, complex mesh?)

---

Pulling out the definition of "geometric objects" would allow to define the same things that are currently supported via the `colliders`, but regardless of their intended _usage_. These "geometric objects" could then be a special "incarnation" of meshes, so to speak. 

Meshes could then be defined as such "geometric objects", roughly like this:
```
    meshes: {
        extensions {
            TODO_geometric_object {
                sphere: {
                    radius: 123
                }
            }
        }
    }
```

These "geometric/parametric meshes" could then be used for rendering **and** as "collider meshes" - roughly like this:
```
    meshes: [
      /* 0 : The high-detail mesh that will be rendered, as a standard glTF mesh */
      /* 1 : A convex hull of mesh 0 (may also be a standard glTF mesh) */
      /* 3 : A geometric object, e.g. a sphere, defined as above */
    ]
    nodes : [

      // A node that renders the high-res mesh and uses 
      // the convex hull as its collider
      {
        mesh : 0,
        ...
          collider : { mesh : 1 }
      }
      
      // A node that renders the high-res mesh and uses 
      // the sphere as its collider
      {
        mesh : 0,
        ...
          collider : { mesh : 2 }
      }

      // A node that renders the sphere (!) and also uses
      // the sphere as its collider
      {
        mesh : 2,
        ...
          collider : { mesh : 2 }
      }
   ]
```

**Comment by eoineoineoin** [comment]
Not opposed to any kinds of renames at all - "rigid bodies" would certainly be better than "physics" in that regard.

However, I'm giving a bunch of thought to what a split would look like. I'm not sure my opinion is _fully_ formed here, but here's sort of a brain-dump of my current thinking.

It was very useful to hear that there was interest in use-cases for simply specifying collision volumes. From our perspective, I'd be reluctant to specify our new primitives (sphere, capsule, hull, etc.) as a type which could be rendered - the existing glTF specification is focussed on triangles; allowing the ability to render spheres would add some complexity, which I think we'd see some push back on? I imagine that those primitives would need to mandate a specific tesselation algorithm (I'd worry about assigning UVs, for example) and things like how those primitives would work with morph targets, skinning, etc. my own background is much more in physics than in graphics, so I don't feel I have sufficient authority to weigh in here. Not sure if you have strong opinions here, Marco?

But, suppose we split it up into a "KHR_collision" and "KHR_rigidbody" spec. It really would be super to enable these additional use-cases (say, for 3D UI) which didn't need simulation. For a single application, you could load a set of assets which contained only KHR_collision nodes and get appropriate behaviour. However, for a more complex application which contained both a rigid body simuation as well as a 3D UI, we'd need some method to distinguish the intended _usage_ of a KHR_collision node.

To support that, I think that the KHR_collision probably wouldn't have any concrete use (outside of applications with a very narrow scope) and the KHR_rigidbody specification would need to refer to KHR_collision and add some additional fields for physics colliision response which wouldn't be applicable to other use-cases. e.g., the schema would read something like:

KHR_rigidbody.physics.collider.schema.json
```
{
    "title" : "KHR_physics glTF extension",
    "type" : "object",
    "$ref" : "node.HKR_collision.collider.schema.json"
    "description" : "Parameters describing a physics collision geometry."
    "properties" : {
        "material": {
        "rigidbodyMaterial" : {
            "allOf": [ { "$ref": "glTFid.schema.json" } ],
            "description": "The index of the physics material in the top level rigidbodyMaterials array."
        },
...
```
i.e., the KHR_rigidbody declares a type of "collider" object which inherits the properties of KHR_collider and adds some additional ones specific to physics simulation.

The other alternative would be to add a "collider" field of type KHR_collision.collider to "rigid_body," but I'd be reluctant to do that, as I really do like the "implicit static" behaviour we get with the spec as-is. It reduces 

A 3D UI "collision" volume would need something similar, inheriting from KHR_collider. This feels a _little_ clunky, but I don't see any other way we could distinguish the intended usage of those colliders. Maybe this is acceptable? When designing "MSFT_Physics", we worked really hard to minimize the amount of types and parameters they had, which is why capsules, colliders and joints use an additional node in the tree to specify a basis for those types.

> (The fact that the convex refers to a mesh raises the question: How to ensure that this mesh actually is a convex hull, and not a "real", arbitrarily-shaped, complex mesh?)

I don't think this is a technical issue (maybe just documentation) - I don't think we should make such a requirement that the mesh be convex. It would be very hard to enforce, and probably just frustrate content creators. We used a reference to a `mesh` in order to minimize the amount of additional types that were required in the specification and the collision geometry that you get out should be a convex hull which _encloses_ the mesh, which I think is alright; a convex hull with vertices {a, b, c, d} is by definition equivalent to a hull {a, b, c, d, (a+b+c+d) * 0.25} in terms of the volume they represent.

In the demo I showed, the two highlighted objects refer to the same (concave) mesh, and it's the responsibility of the physics simulation to generate the hull (and again, I don't want to mandate a particular algorithm here, but physics engines will typically use a well-known algorithm (e.g. quickull) to calculate a minimal convex hull:

![2022-11-25 13_11_46-Havok-Hulls](https://user-images.githubusercontent.com/1676295/203993818-2d01ed36-a0f5-454f-8aae-ce7d8b7a3ab5.png)

**Comment by bhouston** [comment]
@eoineoineoin wrote:
> Not opposed to any kinds of renames at all - "rigid bodies" would certainly be better than "physics" in that regard.

Given there are a lot of types of physics, this makes a lot of sense.

@javagl wrote:
> I think that splitting the geometry definition and its usage could make sense.

@eoineoineoin wrote:
> From our perspective, I'd be reluctant to specify our new primitives (sphere, capsule, hull, etc.) as a type which could be rendered

I am not completely sold on having uniform primitives for rendering and also using them as collision geometry.  I understand the reasoning, but they are so rarely used and it would require a bunch of work to do adaptive tessellation based on view that I am not sure it is worth it.  Personally I am not in favour of this split to have renderable geometry primitives.

@eoineoineoin wrote:
> To support that, I think that the KHR_collision probably wouldn't have any concrete use (outside of applications with a very narrow scope)

I think it would be useful in KHR_behavior in allowing people to specify collision geometry for when you can click on a node or interactive with it in many different ways.

Here is an example of KHR_behavior's behavior graph in action - click on the node to see things start to move - KHR_collision could make this more efficient:

https://interx.on.fleek.co

The reason why I do not want to say use KHR_rigid_body/physics is because maybe you don't actually have a physics engine and thus you can not support all of that extension, but you do support clicking on nodes.

On the web, people click on things constantly.  There is a lot less things to do with physics.

@eoineoineoin wrote:
> the KHR_rigidbody declares a type of "collider" object which inherits the properties of KHR_collider and adds some additional ones specific to physics simulation.

I am not sure it has to derive from it, because that would mean things that could use the KHR_collider objects would have to look for KHR_rigid_bodies as well.   Instead you could have KHR_rigid_body require KHR_collider.  Basically KHR_rigid_body uses KHR_collidier information in its simulation.

BTW this isn't new.  Another group came up with this idea independently: 
https://github.com/omigroup/gltf-extensions/blob/15e397a24c93bff3448d797eecacb9fde6b0177d/extensions/2.0/OMI_collider/README.md.  Although I think that whether or not a collider is a trigger is something that is defined externally to the collider specification.

**Comment by eoineoineoin** [comment]
Hey Ben,

Don't get me wrong - I'm very much willing to seperate the collision from rigid bodies, but I think there's a few details which would need to be worked out. To me, it feels like having a "collider" node by itself isn't sufficient for a content creator to describe the intended usage; there probably needs to be another object to describe the semantics of what that collider represents.

I have two issues in-mind:

Right now, for dynamic bodies, we say "their collision geometry is the union of all child colliders" (consider a car, which would be a dynamic body, with cylindrical colliders for wheels and convex colliders to represent the chassis) - this is also the same strategy that your linked OMI_collider uses. However, it's concievable that if colliders were used for UI nodes as well, you might want to parent a UI node underneath a rigid body, so that the UI moves with the object that it's associated with. I don't know how in this scenario, we can distinguish between a collider which is used for collision detection versus UI. One approach might be to make a new collider definition for a "compound" which has an array of child node indices, explicitly declaring a compound collision geometry - I don't really like this, as it feels like it's a poor-man's implementation of the existing scene tree.

Secondly, there's the "physics material" properties; obviously, we want to be able to specify a different material for wheels versus a chassis; these properties would not belong in the KHR_colliders spec, since they're not applicable to non-physics use-cases. To me, that implies that KHR_rigidbody would add an object which sits alongside the collider; the existance of which would imply that the geometry should generate a physical response (and therefore, the collider be incorporated into the physics simulation) - this would still need some careful consideration; I'd really like to be able to specify a material in such a way that it was still inherited by child colliders. e.g.:

```
WineGlass {RigidBody, Material(Glass)}
     Base {Collider}
     Stem {Collider}
     Bowl {Collider}
```

But a scenario like that would imply that nodes intended for UI would go into the physics sim, which, due to the first issue, I think is undesirable.

Eoin

**Comment by eoineoineoin** [comment]
I'll also note that the [USD Physics](https://graphics.pixar.com/usd/release/api/usd_physics_page_front.html) specification doesn't have any "free" colliders. Personally, I'd really like to be able to keep the "USD Physics" as similar as possible to the "glTF Physics," so will mention this to the USD group and see if I can get some additional viewpoints here.

**Comment by javagl** [comment]
> (@eoineoineoin) Not sure if you have strong opinions here, Marco?

> (@bhouston) I am not completely sold on having uniform primitives for rendering and also using them as collision geometry. 

What I wrote there was not intended as a "proposal" of any sorts. 

I just noticed the overlap: 

- Some people want to _render_ spheres, boxes and cylinders, and define JSON structures like `sphere: { radius: 123 }` or `box: { sizes: [1,2,3] }` or `cylinder: { height: 123, radius 45 }`. 
- Some people want to use spheres, boxes and cylinders as rigid bodies, and essentially define  _the same_ JSON structures
- Some people want to use spheres, boxes and <strike>cylinders</strike> capsules as collision primitives, and essentially define  _the same_ JSON structures...

I don't know enough about rendering, simulation, or collision detection to propose any "solution" here. And I **am** aware of many of the specifics that come into play in each area. One example is the point that you just discussed, namely "How to organize material information (and associate it with the objects)". In the area of rendering, different questions arise - mainly that of tesselation or UV mapping. 

I think that _at the core_, having a specification for the description of geometric objects could make sense. And one could handwavingly come up with ideas of how to "augment" these objects with the required information:
```
meshes: [
    { 
        "geometricObject:" { "sphere" { "radius": 123 }}},
        "tesselation" { "detail" : 100 }
        "uvMapping" { ... }
    }
],
colliders: [
    { 
        "geometricObject:" { "sphere" { "radius": 123 }}},
        "material" { "elasticity" : 0.8 }
    }
],
rigidBody: [
    { 
        "geometricObject:" { "sphere" { "radius": 123 }}},
        "bodyProperties" { "mass" : 12.34 }
    }
]
```
But as you see, this is only a _very_ high-level idea, based on my experience: I hardly ever thought: "These _two_ pieces should have been _one_ piece". But I occasionally thought "This _one_ piece should have been split into _two_ pieces". Small, self-contained, client-specific 'things' (specifications, interfaces, structures) tend to be more versatile, easier to maintain, and longer-lasting.

**But...** ... I say all this with the huge disclaimer: Even though my gut feeling is that _not_ defining such a fundamental core like a "parametric object description" extension _will_ bite back at some point, I cannot profoundly _propose_ that it should be done that way. It might well be that the requirements or caveats in the areas of rendering/simulation/collision detection are _so_ specific for the respective field that trying to unify them does not make sense. 

That's up for you (and the group) to decide.

**Comment by aaronfranke** [comment]
For some inspiration of how this split would work, I'd recommend taking a look at the [`OMI_collider`](https://github.com/aaronfranke/gltf-extensions/tree/OMI_collider/extensions/2.0/OMI_collider) and [`OMI_physics_body`](https://github.com/aaronfranke/gltf-extensions/tree/OMI_physics_body/extensions/2.0/OMI_physics_body) extensions, which are already split as this proposal describes. OMI_collider has already been mentioned in this thread, but I figured it would be helpful to link to the latest iterations of the specs.

> @eoineoineoin: i.e., the KHR_rigidbody declares a type of "collider" object which inherits the properties of KHR_collider and adds some additional ones specific to physics simulation.
> 
> The other alternative would be to add a "collider" field of type KHR_collision.collider to "rigid_body,"

Instead of either of these things, I would recommend instead having KHR_collision be defined on child glTF nodes. This is a simple solution that also automatically gives us the ability to move the shape relative to the body's origin. This is already how OMI physics works, OMI_physics_body must have OMI_collider children to function.

> It was very useful to hear that there was interest in use-cases for simply specifying collision volumes.

For the shapes as generic volumes use case, note that OMI allows specifying being a trigger on either the collider or on the physics body. This is for primarily to enhance compatibility with different game engines, but it also helps with this use case. A volume can be specified as a trigger to indicate you want to use it for non-collision purposes. For example, if you want a specific volume of space to detect things, or to have different audio reverberation properties, or custom gravity (think Super Mario Galaxy), you can specify that shape as a trigger.

@eoineoineoin mentioned that KHR_collision wouldn't have a concrete use, but at least with OMI_collider, there are definitely concrete uses. A non-trigger OMI_collider without a body can be used for static level geometry, which simplifies the implementation in tools that don't want to support a full physics system but do want to support editing level geometry. The other use case is as a generic trigger volume as mentioned in the above paragraph.

I would not worry about rendering these shapes. If somebody wants to do that, they can add another extension that would take a collider shape and render it - there's no need for discussing putting it in this spec.

**Comment by eoineoineoin** [comment]
Hey all,

So, I was experimenting with this a little on a branch (https://github.com/eoineoineoin/glTF_Physics/tree/splitextension) - just wondering what people's thoughts on possible ways to link the "collider" spec to the "rigid bodies" spec would be. I think it's a choice of three options, each with positives/negatives. Without paying attention to the naming of the members (yet), I think they are:

# Option 1

Let the collider and rigidbody sit beside each other like so:

![image](https://user-images.githubusercontent.com/1676295/212094184-a1fc7a7d-1b3f-4d93-8c70-d05e026ffec7.png)

Idea here is that you can add colliders to individual nodes. By themselves, they wouldn't do anything, but if that same node had a "KHR_RigidBodies" object, we could use the existence of some property (in this example, it's "physicsMaterial") to indicate that the extension responsible for importing RigidBodies should look for the KHR_CollisionGeometry object on the same node. Feels like triggers would tie nicely into this.

This has the nice property that you can add a "MyUIControl" extension to the same node and it would pick up the same collider, without duplication. However, it implies that two (otherwise independent) extensions which require collision geometry need to use the same collision geometry for their disparate representations, which I think is not always desirable (e.g., I think UIs would probably want to use boxes, rather than convex hulls)

#2 Option 2

The `node.KHR_RigidBodies.rigidBody.collider` field would have a type of `node.KHR_CollisionGeometry` (i.e., it would just cross-reference the colliders schema:

![image](https://user-images.githubusercontent.com/1676295/212095960-c863a1cc-8228-4a33-852e-79c8223f86b0.png)

This _does_ allow you to use different colliders for different extensions, which I think makes sense, as each extension might have slightly different use-cases. It does lead to potential duplication, if you *do* want the same representation, but I think I'd err on the side of more flexibility, at the cost of some extra bytes.

#3 Option 3

Follow the OMI_Physics approach (https://github.com/aaronfranke/gltf-extensions/tree/OMI_physics_body/extensions/2.0/OMI_physics_body) where the scene can have an array of colliders, which RigidBodies/UI/etc. could index into. This could potentially minimize duplication in theory, but in practice, I think it adds a horizontal dependency between extensions which should otherwise be distinct -- i.e., you have multiple extensions writing to that array and the "UI extension" would need to know how the "rigid body extension" exported the collider and would need to check for equivalence - I don't know how an extension could do this portably without knowing all the details of all the other extensions in use. Of course, if the duplication isn't a problem, I'm sure we could handle multiple extensions writing to the scene's colliders array (have not tried to write an exporter for this strategy yet.)


Right now, I'm leaning towards option 2, but wonder if there's any additional insights here?

**Comment by aaronfranke** [comment]
@eoineoineoin With options 1 and 2, how are you supposed to add multiple primitive collision shapes to one body? I would recommend the way it be done is that colliders be put on child nodes. Child nodes also allow a shape to be offset and/or rotated relative to the parent body. In fact, there is a strong argument to make that be the only option, this is what we have done in OMI_collider. Especially with option 2, there's not really a clear benefit to allowing it on the same node as the physics body.

> would need to know how the "rigid body extension" exported the collider and would need to check for equivalence - I don't know how an extension could do this portably without knowing all the details of all the other extensions in use.

This is not a problem in OMI_collider. Only the collider extension directly deals with the document-level list of colliders, and the OMI_physics_body extension uses child colliders.

The de-duplication aspect is useful for any engine that has shape resources (such as Godot) so that you don't create more resources than needed. However, this de-duplication could also be done by an importer at runtime, so technically this is not important to have in the file level.

**Comment by eoineoineoin** [comment]
> With options 1 and 2, how are you supposed to add multiple primitive collision shapes to one body?

Haven't changed the mechanism around how this works - I agree that instancing and transforming colliders should be done via child nodes. There was a desire to allow multiple primitives per collider (for trimesh and convex types) here, but again, I don't think that'll fundamentally change how the document is structured https://github.com/eoineoineoin/glTF_Physics/issues/11#issue-1508718129

> Only the collider extension directly deals with the document-level list of colliders, and the OMI_physics_body extension uses child colliders.

I actually think this is a defect and that it would prevent usage of the extension outside of simple examples. If you did go down this route, the OMI_Collider exporter extension would need to know how to inspect any possible editor object to extract the collider details. That means, a user can't add an editor object which contains a collider, as the OMI_Collider object won't know where to look for it. The responsibility for exporting colliders *must* be left with the extension which actually makes use of that collider (i.e. the RigidBody extension needs to be able to write into the Collider extension objects.)

**Comment by aaronfranke** [comment]
@eoineoineoin I don't understand why a RigidBody extension would need to write into the collider extension directly. A RigidBody must have a collider shape to have any function, and that shape would be written by the exporter. I also don't know what you mean by editor object.

**Comment by eoineoineoin** [comment]
Sorry, let me be more clear with an example. Suppose you had two seperate exporters for OMI_Collision and OMI_RigidBody. OMI_Collision will write colliders, while OMI_RigidBody will write rigid body data.

In Blender, the Blender object has a field "rigid_body", which has a pointer to a collider that Blender uses for collision geometry. The OMI_Collision exporter would have to inspect the (Blender) rigid_body. If a user added a field to the Blender object, say "my_3d_button_widget" which also had a pointer to a collider, OMI_Collision would have to know look inside that, too! Now, OMI_Collision has a hidden dependency on the "OMI_RigidBody" and "OMI_My_3D_Widget" extension, when, in fact, the dependency should be the other way around. This obviously doesn't work when you consider that users can have arbitrary numbers of niche/unpublished extensions.

> A RigidBody must have a collider shape to have any function

I know it's not the main topic of discussion, but this is actually not quite right; a rigid body only needs mass properties. This is useful when you have constrained systems of joints - you can simulate (e.g.) swinging lights and don't need to add any additional geometry to the simulation.

**Comment by aaronfranke** [comment]
The dependency is the other way around, "OMI_RigidBody" and "OMI_My_3D_Widget" depend on the collision.

The best way I can see to go about this is to make the exporters for "OMI_RigidBody" and "OMI_My_3D_Widget" pass along the collision pointer to the part that exports OMI_Collision. Then there's only one part of the code that handles the responsibility of writing collision shapes to the glTF file.

**Comment by javagl** [comment]
I'm not deeply involved in the details of the recent discussion here. But looking at "Option 2", which defines
```
"collider": {
    "pyhsicsMaterial": 1,
    `"box": { "size": [2,2,2] }
}
```
I have to ask a question that is similar to the gist of my fist comment here: Why is the `box` inlined, and not an index? Whatever the answer is, there's the follow-up question: Why is the `physicsMaterial` defined with an index, but the shape (`box`) is _not_ defined with an index? 

One aspect that may be relevant here is that using an _index_ could reduce redundancy. Consider
```
"collider": {
    "pyhsicsMaterial": 0,
    "box": { "size": [2,2,2] }
}
...
"collider": {
    "pyhsicsMaterial": 1,
    "box": { "size": [2,2,2] }
}
...
"collider": {
    "pyhsicsMaterial": 2,
    "box": { "size": [2,2,2] }
}
...

```
versus
```
"shapes": [ {
    `"box": { "size": [2,2,2] }
} ],
...
"collider": {
    "pyhsicsMaterial": 0,
    "shape": 0
}
...
"collider": {
    "pyhsicsMaterial": 1,
    "shape": 0
}
...
"collider": {
    "pyhsicsMaterial": 2,
    "shape": 0
}
...

```
It would allow modeling the concept of _identity_ cleanly (which might be related to some of the recent discussion - again, I have to read that more thoroughly)

An aside: This is unrelated to the question of whether "shapes" should be extracted into their own `TODO_geometric_shapes` extension. The question of whether something is repeatedly inlined or referred to via an index also applies to the elements _within_ one extension, with the same implications (of being able to avoid redundancy and model _identity_).

**Comment by eoineoineoin** [comment]
From the discussion we had in the 3D formats meeting last week, there's divided opinions on this, but I think Marco is correct - while it's true that it adds complexity to the importers/exporters and in practice, I rarely see the same shapes re-used on different nodes (at least in the AAA games industry,) it's important for people who _do_ need that functionality. I'll re-shuffle the extension appropriately.

**Comment by aidinabedi** [comment]
Hello @eoineoineoin! Having thought significantly more about this, I completely agree. Reusable colliders opens up much more interesting possibilities and is extremely valuable for those that need it.

In fact, having discussed internally - a natural need is reuse of compound colliders like that in bullet, physx, and Unreal :). Unlike a node tree, a compound collider would just be simple array of transforms and shapes - and importantly it would be possible to reuse complex collision volumes made out of simple primitives that are naturally configured to not collide internally (without the need for masks). It would mirror the array of primitives in gltf's mesh object and also fit perfectly with the APIs I've reviewed in physx and bullet. In Unreal, we actually tie the compound collider to the mesh so it would be a big use there as well.

**Comment by aaronfranke** [comment]
@aidinabedi I am against compound colliders, it would make things more complicated for little benefit.

I would prefer that if you need multiple collider shapes, you add each as a collider node, with each node only having one shape. Then these can be combined under the body, and physics engines can optionally treat multiple shapes in one body as a compound shape if that's supported. This is also why I'm in favor of a static bodies (via a string enum instead of this repo's `isKinematic` flag), so you can combine multiple shapes under one body and avoid them colliding internally etc.

Basically, make collider shape nodes, and the only thing that can combine them is a body.

**Comment by eoineoineoin** [comment]
I'm inclined to make joints a document-level property, too, just to make the whole spec consistent with respect to component re-use. i.e., instead of specifying the limits in the node, it indexes into the document:

```
        {
            "name" : "jointSpaceA",
            "rotation" : [....],
            "translation": [...]
            "extensions" : {
                "MSFT_RigidBodies" : {
                    "joint" : {
                        "connectedNode" : 0,
                        "enableCollision" : false,
                        "constraintLimits:" 0               <- indexes into an array in the document
                    }
                }
            },
        },
```

Where the document would have extension data like:

```
"extensions" : {
        "MSFT_RigidBodies" : {
               "physicsJoints": [
                     [ 
                            { "linearAxes" : [0, 1, 2 ] }
                            { "angularAxes" : [1], min: -0.78, max: 0.78 }
                     ]                                                 <- A door hinge
               ]
      }
}
```

Any objections?
